server/channels/app/notification_push.go

```diff
@@ -469,16 +496,31 @@ func (a *App) SendAckToPushProxy(ack *model.PushNotificationAck) error {
 		return nil
 	}
 
-	a.NotificationsLog().Info("Notification received",
-		mlog.String("ackId", ack.Id),
-		mlog.String("type", ack.NotificationType),
-		mlog.String("deviceType", ack.ClientPlatform),
-		mlog.Int("receivedAt", ack.ClientReceivedAt),
+	a.NotificationsLog().Trace("Notification successfully received",
+		mlog.String("type", TypePush),
+		mlog.String("ack_id", ack.Id),
+		mlog.String("push_type", ack.NotificationType),
+		mlog.String("post_id", ack.PostId),
+		mlog.String("ack_type", ack.NotificationType),
+		mlog.String("device_type", ack.ClientPlatform),
+		mlog.Int("received_at", ack.ClientReceivedAt),
 		mlog.String("status", model.PushReceived),
 	)
 
 	ackJSON, err := json.Marshal(ack)
 	if err != nil {
+		a.NotificationsLog().Error("Failed to parse JSON for push proxy ack",
+			mlog.String("type", TypePush),
+			mlog.String("status", StatusServerError),
+			mlog.String("reason", ReasonServerError),
+			mlog.String("ack_id", ack.Id),
+			mlog.String("push_type", ack.NotificationType),
+			mlog.String("post_id", ack.PostId),
+			mlog.String("ack_type", ack.NotificationType),
+			mlog.String("device_type", ack.ClientPlatform),
+			mlog.Int("received_at", ack.ClientReceivedAt),
+			mlog.Err(err),
+		)
 		return fmt.Errorf("failed to encode to JSON: %w", err)
```

Comment:
Returning an error and logging at the same place is a bad pattern. There's also a lot of repetition here. We can do much better:
1. Create a new logger out of the existing logger by adding those fields which are repeated everwhere. For example:

```
c.Logger = c.App.Log().With(
		mlog.String("path", c.AppContext.Path()),
		mlog.String("request_id", c.AppContext.RequestId()),
		mlog.String("ip_addr", c.AppContext.IPAddress()),
		mlog.String("user_id", c.AppContext.Session().UserId),
		mlog.String("method", r.Method),
	)
```

2. You already have the `ack` struct at the call site. Any reason we can't add some more context to the error and improve the logging there?

Reply:
>  There is some repetition as you say but it may not be enough to justify creating a separate object.

It seemed to me that there is some repetition. We can use this to refactor the common fields, and then add the other remaining fields during logging.

> Speaking of patterns, can you elaborate on why returning an error and logging is a bad pattern? 

Because it complicates testing. You cannot really test whether you log something or not (until you mock the logger and test the output), but you can easily test what a function returns. This is also the reason why logging in the store layer is not a good idea.

Additionally, it seems to me that a lot of the information that you are logging is already there in the call site. So what you really want is the additional contextual information from the error with the additional fields.

>  but when we're trying to be more verbose, and a method returns different errors

Ideally a method shouldn't return different errors. We can have a custom error type and add all the necessary details. Being verbose is completely fine, but I think we can keep the code clean and be verbose at the same time :)

Reply:
> It seemed to me that there is some repetition. We can use this to refactor the common fields, and then add the other remaining fields during logging.

So would we create a logger object anywhere we end up using one or more of the same fields within the same code block? Seems a bit heavy handed to me but if that's the preferred method I'm okay with that.

> Because it complicates testing. You cannot really test whether you log something or not (until you mock the logger and test the output), but you can easily test what a function returns. This is also the reason why logging in the store layer is not a good idea.

Gotcha. This might be a hot take, but I've never seen an incredible amount of value in testing whether something was logged or not. Logging is a bit of a side-effect thing to me in that we need to note that the app behaved a certain way, and that doesn't really affect the logic of a method, at least from a unit perspective. I could see some bigger E2E tests around certain logging behaviours, but nothing where we'd want to explicitly test that logging happens.

Just my 2 cents. If we're talking about mocking though then that can be complicated. In the Desktop App for example we have a global mock for our logging library that just ignores it all when testing.

> Ideally a method shouldn't return different errors. We can have a custom error type and add all the necessary details

So the preference here would be to instead of logging in the method itself would be to return all the data we want in error objects? Is that something we already do a lot of, or is that something we want to start doing? I'm all for it, but I don't see a lot of that happening right now so it might be a pattern we want to change later.